File: scripts/monthly-ingestion.ts
// ---------------------------------------------------------------------------
//  Monthly ingestion of IRDAI Non-Life downloadable Excel sheets.
//
//  1. Ask the muns chat agent to scrape https://irdai.gov.in/non-life and
//     return every downloadable Excel link.
//  2. Pull every IRDAI Excel download URL out of the agent's answer.
//  3. Download only the links we have NOT already saved (deduped via a
//     manifest), into the repo-root folder `non-life-monthly/`.
//
//  Downloads go through the repo's shared fetchBuffer(), which chains three
//  tiers to get past IRDAI's WAF: browser-headered fetch -> headless Chromium
//  -> an optional fetch proxy (INGEST_FETCH_PROXY). IRDAI returns 403 to plain
//  requests from GitHub's datacenter IPs, so those fallbacks are what make the
//  download actually land.
//
//  Dedup rule: we key on the full download URL. IRDAI bumps the version/
//  timestamp in the URL whenever a sheet is revised, so a revised sheet reads
//  as a *new* link and is saved as a new file; an unchanged link is skipped.
//
//  The access token is read from MUNS_API_TOKEN (a GitHub Actions secret).
// ---------------------------------------------------------------------------

import { mkdir, readFile, writeFile, access } from 'node:fs/promises'
import { dirname, resolve, extname, basename } from 'node:path'
import { fileURLToPath } from 'node:url'
import { fetchBuffer } from './ingest/parsers'

const HERE = dirname(fileURLToPath(import.meta.url))
const REPO_ROOT = resolve(HERE, '..')
const OUT_DIR = resolve(REPO_ROOT, 'non-life-monthly')
const MANIFEST_PATH = resolve(OUT_DIR, 'manifest.json')

const API_URL = 'https://devde.muns.io/chat/chat-muns'
const TARGET_PAGE = 'https://irdai.gov.in/non-life'
const API_TIMEOUT_MS = 600_000 // the agent scrapes the page live; give it room.

// Any IRDAI document download URL, stopping at whitespace / markdown pipes /
// quotes / angle brackets / closing parens.
const URL_RE = /https:\/\/irdai\.gov\.in\/documents\/[^\s)|"'<>]+download=true/g

interface ManifestEntry {
  filename: string
  bytes: number
  fetched_at: string
}
interface Manifest {
  files: Record<string, ManifestEntry>
  updated_at?: string
}

function buildPayload() {
  // The scrape window rolls with the run date so a recurring job always asks
  // over a current 2-year window.
  const today = new Date()
  const from = new Date(today)
  from.setFullYear(from.getFullYear() - 2)
  const iso = (d: Date) => d.toISOString().slice(0, 10)
  return {
    user_index: 124,
    tasks: [
      'scrape and gimme a link of all downoadable excel sheets on ' +
        'https://irdai.gov.in/non-life',
    ],
    query_context: {
      TICKER_SYMBOL: [],
      FROM_DATE: iso(from),
      TO_DATE: iso(today),
      ANNOUNCEMENT_FORM_TYPE: 'all',
      DOCUMENT_IDS: [],
      CATEGORIES: [],
      WEB_SEARCH_ENABLED: true,
      COUNTRY: [],
      CONTEXT_EMAIL: 'nadamsaluja@gmail.com',
      CONTEXT_COMPANY_NAME: [],
      GET_ANNOUNCEMENTS_ENABLED: false,
      chatHistory: [],
      mode: 'fast',
    },
    autoAddUpcoming: false,
    urls: [],
  }
}

async function callAgent(token: string): Promise<string> {
  const ctrl = new AbortController()
  const timer = setTimeout(() => ctrl.abort(), API_TIMEOUT_MS)
  try {
    const res = await fetch(API_URL, {
      method: 'POST',
      headers: {
        accept: '*/*',
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(buildPayload()),
      signal: ctrl.signal,
    })
    if (!res.ok) {
      throw new Error(`agent call failed: HTTP ${res.status} ${res.statusText}`)
    }
    return await res.text()
  } finally {
    clearTimeout(timer)
  }
}

/** The <ans>...</ans> section (the curated, deduped list) if present, else all. */
function extractAnswer(text: string): string {
  const m = text.match(/<ans>([\s\S]*?)<\/ans>/)
  return m ? m[1] : text
}

function urlsIn(text: string): string[] {
  const seen = new Set<string>()
  const out: string[] = []
  for (const m of text.matchAll(URL_RE)) {
    const u = m[0]
    if (!seen.has(u)) {
      seen.add(u)
      out.push(u)
    }
  }
  return out
}

/**
 * All IRDAI Excel download URLs from the response, de-duplicated, order
 * preserved. Prefer the curated <ans> list; if that has none (the agent
 * sometimes answers without a clean table), fall back to scanning the whole
 * response — the same links also appear in the <sources> block.
 */
function extractUrls(text: string): string[] {
  const fromAns = urlsIn(extractAnswer(text))
  return fromAns.length > 0 ? fromAns : urlsIn(text)
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms))

/**
 * Call the agent until it returns at least one link. The agent is an LLM and
 * occasionally answers without actually scraping, so we retry a few times
 * before giving up.
 */
async function getLinks(token: string, attempts = 3): Promise<string[]> {
  let urls: string[] = []
  for (let i = 1; i <= attempts; i++) {
    console.log(`Calling agent to scrape ${TARGET_PAGE} (attempt ${i}/${attempts}) ...`)
    const raw = await callAgent(token)
    urls = extractUrls(raw)
    if (urls.length > 0) {
      console.log(`Found ${urls.length} distinct download link(s).`)
      return urls
    }
    // Diagnostic: show how the agent answered when it gave us nothing.
    const snippet = raw.replace(/\s+/g, ' ').trim().slice(0, 600)
    console.error(`  no links in response (${raw.length} chars). Head: ${snippet}`)
    if (i < attempts) await sleep(5000)
  }
  return urls
}

/**
 * Build a clean, human-readable .xlsx filename from a download URL. The path
 * segment before the document UUID holds the original filename (URL-encoded,
 * often a Hindi title + ' _ ' + the English title). We decode it, keep up to
 * the first '.xlsx', prefer the English title after ' _ ', and sanitise.
 */
function deriveFilename(url: string): string {
  const parts = new URL(url).pathname.split('/')
  const uuid = parts[parts.length - 1] || ''
  const enc = parts[parts.length - 2] || ''

  let name: string
  try {
    name = decodeURIComponent(enc.replace(/\+/g, ' ')).trim()
  } catch {
    name = enc.replace(/\+/g, ' ').trim()
  }

  const xlsxIdx = name.toLowerCase().indexOf('.xlsx')
  if (xlsxIdx !== -1) name = name.slice(0, xlsxIdx + 5)
  if (name.includes(' _ ')) name = name.split(' _ ').pop()!.trim()

  // Keep word chars, spaces, dash, dot; collapse whitespace.
  name = name.replace(/[^\w\-. ]/g, '').replace(/\s+/g, ' ').trim()

  if (!name.toLowerCase().endsWith('.xlsx')) {
    name = name ? `${name}.xlsx` : `irdai-nonlife-${uuid}.xlsx`
  }
  return name
}

async function exists(p: string): Promise<boolean> {
  try {
    await access(p)
    return true
  } catch {
    return false
  }
}

/** If `name` is taken, suffix with -2, -3, ... */
async function uniqueName(name: string, used: Set<string>): Promise<string> {
  const ext = extname(name)
  const base = basename(name, ext)
  let candidate = name
  let i = 2
  while (used.has(candidate) || (await exists(resolve(OUT_DIR, candidate)))) {
    candidate = `${base}-${i}${ext}`
    i++
  }
  used.add(candidate)
  return candidate
}

async function loadManifest(): Promise<Manifest> {
  try {
    const data = JSON.parse(await readFile(MANIFEST_PATH, 'utf8'))
    if (data && typeof data === 'object' && data.files) return data as Manifest
  } catch {
    /* no manifest yet / unreadable — start fresh */
  }
  return { files: {} }
}

async function saveManifest(manifest: Manifest): Promise<void> {
  manifest.updated_at = new Date().toISOString()
  // Stable key order so diffs stay readable.
  const ordered: Manifest = { files: {}, updated_at: manifest.updated_at }
  for (const k of Object.keys(manifest.files).sort()) ordered.files[k] = manifest.files[k]
  await writeFile(MANIFEST_PATH, JSON.stringify(ordered, null, 2) + '\n', 'utf8')
}

async function main(): Promise<number> {
  const token = (process.env.MUNS_API_TOKEN || '').trim()
  if (!token) {
    console.error(
      'ERROR: MUNS_API_TOKEN is not set. Add it as a GitHub Actions secret named MUNS_API_TOKEN.',
    )
    return 1
  }

  await mkdir(OUT_DIR, { recursive: true })
  const manifest = await loadManifest()
  const known = manifest.files

  let urls: string[]
  try {
    urls = await getLinks(token)
  } catch (err) {
    console.error(`ERROR: ${err instanceof Error ? err.message : String(err)}`)
    return 1
  }

  if (urls.length === 0) {
    console.error(
      'ERROR: the agent returned no IRDAI Excel download links after several ' +
        'attempts. Nothing downloaded. Existing files left untouched.',
    )
    return 1
  }

  const usedNames = new Set<string>(
    Object.values(known).map((v) => v.filename).filter(Boolean),
  )
  let newCount = 0
  let failures = 0

  for (const url of urls) {
    if (known[url]) continue // already have this exact link — skip.
    const name = await uniqueName(deriveFilename(url), usedNames)
    const dest = resolve(OUT_DIR, name)
    try {
      const { buffer } = await fetchBuffer(url)
      if (!buffer || buffer.length === 0) throw new Error('empty body')
      await writeFile(dest, buffer)
      known[url] = {
        filename: name,
        bytes: buffer.length,
        fetched_at: new Date().toISOString(),
      }
      newCount++
      console.log(`  + saved: ${name}  (${buffer.length.toLocaleString()} bytes)`)
    } catch (err) {
      failures++
      usedNames.delete(name)
      console.error(`  ! failed: ${name}  (${err instanceof Error ? err.message : String(err)})`)
    }
  }

  await saveManifest(manifest)

  console.log(
    `\nDone. ${newCount} new file(s) saved; ${urls.length - newCount} already had / skipped. ` +
      `Total tracked: ${Object.keys(known).length}.`,
  )

  // Loud failure: links were found but NONE could be downloaded (typically
  // IRDAI's 403 to GitHub IPs). Fail the run so it doesn't look green while
  // doing nothing — the fix is to set the INGEST_FETCH_PROXY secret.
  if (newCount === 0 && failures > 0) {
    console.error(
      '\nERROR: found links but downloaded nothing — every file was blocked ' +
        '(IRDAI returns 403 to GitHub runner IPs). Set the INGEST_FETCH_PROXY ' +
        'repo secret (a proxy URL template with a {url} placeholder) so downloads ' +
        'leave from a non-blocked IP.',
    )
    return 1
  }
  return 0
}

main().then((code) => {
  process.exitCode = code
})

File: scripts/monthly_ingestion.py
#!/usr/bin/env python3
"""Monthly ingestion of IRDAI Non-Life downloadable Excel sheets.

Flow:
  1. Ask the muns chat agent to scrape https://irdai.gov.in/non-life and
     return every downloadable Excel link.
  2. Pull every IRDAI Excel download URL out of the agent's answer.
  3. Download only the links we have NOT already saved (deduped via a manifest),
     into the repo-root folder `non-life-monthly/`.

Dedup rule: we key on the full download URL. IRDAI bumps the version/timestamp
in the URL whenever a sheet is revised, so a revised sheet reads as a *new*
link and is saved as a new file; an unchanged link is skipped. This matches
"on repeat of files / links, only save the new ones each run."

The access token is read from the MUNS_API_TOKEN environment variable (a GitHub
Actions secret) — never hard-coded here.
"""

import datetime
import json
import os
import re
import sys
import urllib.request
import urllib.error
from urllib.parse import unquote, urlparse

# --- config -----------------------------------------------------------------

API_URL = "https://devde.muns.io/chat/chat-muns"
TARGET_PAGE = "https://irdai.gov.in/non-life"
OUT_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "non-life-monthly")
MANIFEST_PATH = os.path.join(OUT_DIR, "manifest.json")

# Matches any IRDAI document download URL, stopping at whitespace / markdown
# table pipes / quotes / angle brackets / closing parens.
URL_RE = re.compile(r"https://irdai\.gov\.in/documents/[^\s)|\"'<>]+download=true")

API_TIMEOUT = 600  # the agent scrapes the page live; give it room.
DOWNLOAD_TIMEOUT = 120
USER_AGENT = "Mozilla/5.0 (compatible; monthly-ingestion-bot/1.0)"


# --- helpers ----------------------------------------------------------------

def build_payload():
    """Build the chat request. The scrape window rolls with the run date so a
    recurring job always asks over a current 2-year window."""
    today = datetime.date.today()
    from_date = today.replace(year=today.year - 2)
    return {
        "user_index": 124,
        "tasks": [
            "scrape and gimme a link of all downoadable excel sheets on "
            "https://irdai.gov.in/non-life"
        ],
        "query_context": {
            "TICKER_SYMBOL": [],
            "FROM_DATE": from_date.isoformat(),
            "TO_DATE": today.isoformat(),
            "ANNOUNCEMENT_FORM_TYPE": "all",
            "DOCUMENT_IDS": [],
            "CATEGORIES": [],
            "WEB_SEARCH_ENABLED": True,
            "COUNTRY": [],
            "CONTEXT_EMAIL": "nadamsaluja@gmail.com",
            "CONTEXT_COMPANY_NAME": [],
            "GET_ANNOUNCEMENTS_ENABLED": False,
            "chatHistory": [],
            "mode": "fast",
        },
        "autoAddUpcoming": False,
        "urls": [],
    }


def call_agent(token):
    """POST the scrape request and return the raw text response."""
    data = json.dumps(build_payload()).encode("utf-8")
    req = urllib.request.Request(
        API_URL,
        data=data,
        method="POST",
        headers={
            "accept": "*/*",
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json",
        },
    )
    with urllib.request.urlopen(req, timeout=API_TIMEOUT) as resp:
        return resp.read().decode("utf-8", "replace")


def extract_answer(text):
    """Return the <ans>...</ans> section (the curated, deduped list) if present,
    else the whole response."""
    m = re.search(r"<ans>(.*?)</ans>", text, re.S)
    return m.group(1) if m else text


def extract_urls(text):
    """All IRDAI Excel download URLs, de-duplicated, order preserved."""
    seen, out = set(), []
    for u in URL_RE.findall(text):
        if u not in seen:
            seen.add(u)
            out.append(u)
    return out


def derive_filename(url):
    """Build a clean, human-readable .xlsx filename from a download URL.

    The path segment before the document UUID holds the original filename
    (URL-encoded, often a Hindi title + ' _ ' + the English title). We decode
    it, keep up to the first '.xlsx', prefer the English title after ' _ ',
    and sanitise. Falls back to the document UUID if anything looks off."""
    path_parts = urlparse(url).path.split("/")
    uuid = path_parts[-1] if path_parts else ""
    enc = path_parts[-2] if len(path_parts) >= 2 else ""

    name = unquote(enc.replace("+", " ")).strip()

    low = name.lower()
    if ".xlsx" in low:
        name = name[: low.index(".xlsx") + 5]
    if " _ " in name:
        name = name.split(" _ ")[-1].strip()

    # Keep word chars, spaces, dash, dot; collapse whitespace.
    name = re.sub(r"[^\w\-. ]", "", name)
    name = re.sub(r"\s+", " ", name).strip()

    if not name.lower().endswith(".xlsx"):
        name = (name + ".xlsx") if name else f"irdai-nonlife-{uuid}.xlsx"
    return name


def unique_path(name, used):
    """Avoid clobbering: if `name` is taken, suffix with -2, -3, ..."""
    base, ext = os.path.splitext(name)
    candidate, i = name, 2
    while candidate in used or os.path.exists(os.path.join(OUT_DIR, candidate)):
        candidate = f"{base}-{i}{ext}"
        i += 1
    used.add(candidate)
    return candidate


def download(url, dest):
    req = urllib.request.Request(url, headers={"User-Agent": USER_AGENT})
    with urllib.request.urlopen(req, timeout=DOWNLOAD_TIMEOUT) as resp:
        body = resp.read()
    with open(dest, "wb") as f:
        f.write(body)
    return len(body)


def load_manifest():
    if os.path.exists(MANIFEST_PATH):
        try:
            with open(MANIFEST_PATH, encoding="utf-8") as f:
                data = json.load(f)
            if isinstance(data, dict) and isinstance(data.get("files"), dict):
                return data
        except (json.JSONDecodeError, OSError):
            pass
    return {"files": {}}  # url -> {filename, bytes, fetched_at}


def save_manifest(manifest):
    manifest["updated_at"] = datetime.datetime.utcnow().isoformat() + "Z"
    with open(MANIFEST_PATH, "w", encoding="utf-8") as f:
        json.dump(manifest, f, ensure_ascii=False, indent=2, sort_keys=True)


# --- main -------------------------------------------------------------------

def main():
    token = os.environ.get("MUNS_API_TOKEN", "").strip()
    if not token:
        print("ERROR: MUNS_API_TOKEN is not set. Add it as a GitHub Actions "
              "secret named MUNS_API_TOKEN.", file=sys.stderr)
        return 1

    os.makedirs(OUT_DIR, exist_ok=True)
    manifest = load_manifest()
    known = manifest["files"]

    print(f"Calling agent to scrape {TARGET_PAGE} ...")
    try:
        raw = call_agent(token)
    except urllib.error.HTTPError as e:
        print(f"ERROR: agent call failed: HTTP {e.code} {e.reason}", file=sys.stderr)
        return 1
    except (urllib.error.URLError, TimeoutError) as e:
        print(f"ERROR: agent call failed: {e}", file=sys.stderr)
        return 1

    urls = extract_urls(extract_answer(raw))
    if not urls:
        print("WARNING: no IRDAI Excel download links found in the response. "
              "Leaving existing files untouched.", file=sys.stderr)
        # Not a hard failure — nothing to add, nothing to lose.
        return 0

    print(f"Found {len(urls)} distinct download link(s) in the answer.")

    used_names = {v["filename"] for v in known.values() if "filename" in v}
    new_count = 0

    for url in urls:
        if url in known:
            continue  # already have this exact link — skip.
        name = unique_path(derive_filename(url), used_names)
        dest = os.path.join(OUT_DIR, name)
        try:
            size = download(url, dest)
        except Exception as e:  # one bad link must not sink the whole run.
            print(f"  ! failed: {name}  ({e})", file=sys.stderr)
            used_names.discard(name)
            continue
        known[url] = {
            "filename": name,
            "bytes": size,
            "fetched_at": datetime.datetime.utcnow().isoformat() + "Z",
        }
        new_count += 1
        print(f"  + saved: {name}  ({size:,} bytes)")

    save_manifest(manifest)

    print(f"\nDone. {new_count} new file(s) saved; "
          f"{len(urls) - new_count} already had. "
          f"Total tracked: {len(known)}.")
    return 0


if __name__ == "__main__":
    sys.exit(main())
